File: Backtracking/subset-sum.c
// Algorithm: Subset Sum (Backtracking)
// Type: Backtracking / Recursive
// Time: O(2^n)
// Space: O(n)

#include <stdio.h>

void printSubset(int subset[], int size) {
    printf("{ ");
    for (int i = 0; i < size; i++)
        printf("%d ", subset[i]);
    printf("}\n");
}

void subsetSumBacktrack(int arr[], int n, int target, int index, int subset[], int subsetSize) {
    if (target == 0) {
        printSubset(subset, subsetSize);
        return;
    }
    if (target < 0 || index == n)
        return;

    // Include arr[index]
    subset[subsetSize] = arr[index];
    subsetSumBacktrack(arr, n, target - arr[index], index + 1, subset, subsetSize + 1);

    // Exclude arr[index]
    subsetSumBacktrack(arr, n, target, index + 1, subset, subsetSize);
}

// Usage:
// int arr[] = {3, 34, 4, 12, 5, 2};
// int subset[6];
// subsetSumBacktrack(arr, 6, 9, 0, subset, 0);

File: Backtracking/subset-sum.java
// Algorithm: Subset Sum (Backtracking)
// Type: Backtracking / Recursive
// Time: O(2^n)
// Space: O(n)

import java.util.ArrayList;

public class SubsetSumBacktrack {
    public static void subsetSumBacktrack(int[] arr, int target, int index, ArrayList<Integer> current) {
        if (target == 0) {
            System.out.println(current);
            return;
        }
        if (target < 0 || index == arr.length)
            return;

        // Include arr[index]
        current.add(arr[index]);
        subsetSumBacktrack(arr, target - arr[index], index + 1, current);

        // Exclude arr[index]
        current.remove(current.size() - 1);
        subsetSumBacktrack(arr, target, index + 1, current);
    }

    public static void main(String[] args) {
        int[] arr = {3, 34, 4, 12, 5, 2};
        subsetSumBacktrack(arr, 9, 0, new ArrayList<>());
    }
}
